### ISSUE TITLE
clang: error: ptxas command failed with exit code 255 (use -v to see invocation)

### ISSUE THREAD

**Issue by rnburn** [llvm/llvm-project#65403]
Clang is crashing trying to compile cuda code. I'm attaching the "Please attach" files.

For more context, this is from https://github.com/rnburn/rls_bltzr_cd/commit/d39e5ba1d10e30bc2e311de3bb30b37e603fd51c which is a project to transition https://github.com/spaceandtimelabs/blitzar from nvcc to clang.

A build environment for the issue can be set up by running 
```
ci/run_docker.sh
bazel build //....
````

I haven't investigated the error very far, so it could have a simple workaround or clang might not be invoked correctly for cuda code. But I thought I would go ahead and file a bug report since it gives a "PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT" message.

Here's the full error I got
```
ERROR: /src/sxt/field51/base/BUILD:6:17: Compiling sxt/field51/base/byte_conversion.cc failed: (Exit 1): clang failed: error executing command (from target //sxt/field51/base:byte_conversion) /usr/bin/clang -x cu '--cuda-gpu-arch=sm_70' '--cuda-gpu-arch=sm_70' '--cuda-path=/usr/local/cuda' '-frandom-seed=bazel-out/k8-fastbuild/bin/sxt/field51/base/_objs/byte_conversion/byte_conversion.pic.o' ... (remaining 33 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
clang: warning: CUDA version 12.1 is only partially supported [-Wunknown-cuda-version]
ptxas fatal   : Unresolved extern function '_ZN3sxt4f51b6reduceEPmPKm'
clang: error: ptxas command failed with exit code 255 (use -v to see invocation)
Ubuntu clang version 18.0.0 (++20230904042322+7af408e89bd0-1~exp1~20230904042432.1164)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
clang: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/byte_conversion-7267a8.cu
clang: note: diagnostic msg: /tmp/byte_conversion-sm_70-9977f9.cu
clang: note: diagnostic msg: /tmp/byte_conversion-7267a8.sh
clang: note: diagnostic msg:

********************
```

**Comment by rnburn** [comment]
[bug.zip](https://github.com/llvm/llvm-project/files/12527701/bug.zip)

**Comment by Artem-B** [comment]
> ptxas fatal   : Unresolved extern function '_ZN3sxt4f51b6reduceEPmPKm'

This looks like a problem with your source code. By default CUDA compilation of every source file must have all required functions available.

The diagnostics is unfortunately unusual, confusing, and easy to miss. We may want to diagnose such unresolved references in clang itself.

**Comment by rnburn** [comment]
> By default CUDA compilation of every source file must have all required functions available.

The goal was to use separate compilation (https://developer.nvidia.com/blog/separate-compilation-linking-cuda-device-code/) -- is it missing some of the options for that?

**Comment by Artem-B** [comment]
>  is it missing some of the options for that

The problem report didn't have complete command line, so I do not know. Probably.

RDC compilation is somewhat complicated. I do not know the current state of support of it in bazel. I think they did recently add some rules for CUDA compilation. 

If you need to do it yourself, here's what tensorflow does: https://github.com/tensorflow/tensorflow/blob/528ad01d26ae47ef9452bf2ef6fd49e9c0f4f2ba/third_party/nccl/build_defs.bzl.tpl#L265

There's also (experimental) support in clang itself which allows compiling all CUDA sources to GPU object files and link them all at the final executable link phase: https://www.youtube.com/watch?v=4NnzymmQe7k

**Comment by rnburn** [comment]
> RDC compilation is somewhat complicated. I do not know the current state of support of it in bazel. I think they did recently add some rules for CUDA compilation.

That's correct. I'm trying to transition to the official bazel rules with clang, https://github.com/bazel-contrib/rules_cuda, instead these [custom rules](https://github.com/spaceandtimelabs/blitzar/tree/main/third_party/rules_cuda) with nvcc.

That new experimental support looks promising - hopefully, it will be available from the bazel rules sometime soon.

**Comment by Artem-B** [comment]
The rules_cuda do seem to mention the `--cuda-rdc` flag: https://github.com/bazel-contrib/rules_cuda/blob/e7725537c7c5107a47b503c5e36752012c5521f9/cuda/private/toolchain_configs/clang.bzl#L100

You need to make sure that the flag does get passed to the clang compilation.
I'm not sure at which point `device_link` kicks in, but you will want to make sure it does happen, too.

**Comment by rnburn** [comment]
Thank you for the help. Setting `rdc = True` seems to have fixed this issue.

**Comment by arsenm** [comment]
Can this be closed?

**Comment by Sirraide** [comment]
> Can this be closed?

Yeah, I don't think this is really a Clang frontend issue: if you forget to define a function, you get a linker error (or in this case a `ptxas` error); NVCC does the same (https://godbolt.org/z/6obKb4qoc).
